pennylane_lightning/core/src/bindings/BindingsMPI.hpp

```diff
@@ -0,0 +1,437 @@
+// Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file Bindings.hpp
+ * Defines device-agnostic operations to export to Python and other utility
+ * functions interfacing with Pybind11.
+ */
+
+#pragma once
+#include <set>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <vector>
+
+#include <pybind11/complex.h>
+#include <pybind11/functional.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/stl_bind.h>
+
+#include "CPUMemoryModel.hpp" // CPUMemoryModel, getMemoryModel, bestCPUMemoryModel, getAlignment
+#include "JacobianData.hpp"
+#include "Macros.hpp" // CPUArch
+#include "Memory.hpp" // alignedAlloc
+#include "Observables.hpp"
+#include "Util.hpp" // for_each_enum
+
+#ifdef _ENABLE_PLGPU
+#include "AdjointJacobianGPUMPI.hpp"
+#include "JacobianDataMPI.hpp"
+#include "LGPUBindingsMPI.hpp"
+#include "MeasurementsGPUMPI.hpp"
+#include "ObservablesGPUMPI.hpp"
+
+/// @cond DEV
+namespace {
+using namespace Pennylane::LightningGPU;
+using namespace Pennylane::LightningGPU::Algorithms;
+using namespace Pennylane::LightningGPU::Observables;
+using namespace Pennylane::LightningGPU::Measures;
+} // namespace
+  /// @endcond
+
+#else
+
+static_assert(false, "Backend not found.");
+
+#endif
+
+namespace py = pybind11;
+
+namespace Pennylane {
+/**
+ * @brief Register observable classes.
+ *
+ * @tparam StateVectorT
+ * @param m Pybind module
+ */
+template <class StateVectorT> void registerObservablesMPI(py::module_ &m) {
```

Comment:
I think we can find a design that avoids creating a separate bindings-MPI file with everything repeated for all the internals.

I'd like us to create a tech debt story to revisit this and try to find a unified design that alows us to use MPI & non-MPI code with the same scaffolding.

Reply:
Totally agree!